### PR TITLE
Bootstrap openshift e2e tests

### PR DESCRIPTION
**Pull request by akram**
# Changes

Adds an example for e2e tests for OpenShift

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes tests (if functionality changed/added)
- [x] Commit messages follow [commit message best practices](https://github.com/jenkinsci/kubernetes-operator/blob/master/CONTRIBUTING.md#commit-messages)

### CODE REVIEW

File: deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ kind: Jenkins
 metadata:
   annotations:
     jenkins.io/openshift-mode: 'true'
```
Comment (waveywaves) [inline deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml:5]: Can you remove this annotation ?

**Comment by waveywaves** [comment]
Lgtm

File: deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml (hunk 1 above)
Comment (tomaszsek) [reply deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml:5]: I agree with Vibhav, please remove this line.

File: test/e2e/jenkins.go [hunk 2]
```diff
@@ -51,6 +51,28 @@ func getJenkinsMasterPod(t *testing.T, jenkins *v1alpha2.Jenkins) *corev1.Pod {
...
+func getServiceAccount(t *testing.T, jenkins *v1alpha2.Jenkins) *corev1.ServiceAccount {
+	saName := fmt.Sprintf("jenkins-operator-%s", jenkins.Name)
+	sa, err := framework.Global.KubeClient.CoreV1().ServiceAccounts(jenkins.ObjectMeta.Namespace).Get(saName, metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	return sa
+}
+
+type JenkinsSample struct {
```
Comment (tomaszsek) [inline test/e2e/jenkins.go:63]:
```suggestion
type JenkinsE2EConfiguration struct {
```

File: test/e2e/openshift_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	defer showLogsAndCleanup(t, ctx)
+
+	numberOfExecutors := 6
+	numberOfExecutorsEnvName := "NUMBER_OF_EXECUTORS"
+	stringData := make(map[string]string)
+	stringData[numberOfExecutorsEnvName] = fmt.Sprintf("%d", numberOfExecutors)
+
+	// base
+	createUserConfigurationSecret(t, namespace, stringData)
+	createUserConfigurationConfigMap(t, namespace, numberOfExecutorsEnvName, "")
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:26]:
```suggestion
```
It's not needed in this test.

File: test/e2e/openshift_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	sample := JenkinsSample{
+		name:              openshfitE2e,
+		namespace:         namespace,
+		priorityClassName: "",
+		seedJob:           &[]v1alpha2.SeedJob{},
+		groovyScripts:     v1alpha2.GroovyScripts{},
+		casc:              v1alpha2.ConfigurationAsCode{},
+	}
+	jenkins := createJenkinsCRFromSample(t, sample)
+	createDefaultLimitsForContainersInNamespace(t, namespace)
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:36]:
```suggestion
	createDefaultLimitsForContainersInNamespace(t, namespace)
```
It's not needed in this test.

File: test/e2e/openshift_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	waitForJenkinsBaseConfigurationToComplete(t, jenkins)
+	verifyJenkinsMasterPodAttributes(t, jenkins)
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:38]:
```suggestion
	verifyJenkinsMasterPodAttributes(t, jenkins)
```
It's not needed in this test.

File: test/e2e/openshift_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	verifyServiceAccountAnnotations(t, jenkins)
+	jenkinsClient, cleanUpFunc := verifyJenkinsAPIConnection(t, jenkins, namespace)
+	defer cleanUpFunc()
+	verifyPlugins(t, jenkinsClient, jenkins)
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:42]:
```suggestion
```
It's not needed in this test.

File: test/e2e/openshift_test.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+	sample := JenkinsSample{
+		name:              "openshift-plugins-e2e",
+		namespace:         namespace,
+		priorityClassName: "",
+		seedJob:           &[]v1alpha2.SeedJob{},
+		groovyScripts:     v1alpha2.GroovyScripts{},
+		casc:              v1alpha2.ConfigurationAsCode{},
+	}
+	createJenkinsCRFromSample(t, sample)
+}
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:68]: For me, this test do nothing useful. What is purpose for this test?

File: test/e2e/jenkins.go (hunk 2 above)
Comment (akram) [reply test/e2e/jenkins.go]: solved

File: test/e2e/openshift_test.go (hunk 4 above)
Comment (akram) [reply test/e2e/openshift_test.go]: removed

File: test/e2e/openshift_test.go (hunk 7 above)
Comment (akram) [reply test/e2e/openshift_test.go]: I rewrote the test to ensure that the passed plugins are present in the instantiated Jenkins

File: test/e2e/openshift_test.go (hunk 6 above)
Comment (akram) [reply test/e2e/openshift_test.go:38]: we need the jenkinsClient to chech that the requested plugins are preset

File: deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml [hunk 8]
```diff
@@ -1,9 +1,7 @@
 apiVersion: jenkins.io/v1alpha2
 kind: Jenkins
 metadata:
-  annotations:
-    jenkins.io/openshift-mode: 'true'
-  name: jenkins
+  name: jay
```
Comment (tomaszsek) [inline deploy/crds/openshift_jenkins_v1alpha2_jenkins_cr.yaml:4]:
```suggestion
  name: jenkins
```

File: test/e2e/jenkins.go [hunk 9]
```diff
@@ -51,12 +51,39 @@ func getJenkinsMasterPod(t *testing.T, jenkins *v1alpha2.Jenkins) *corev1.Pod {
...
+	name              string
+	namespace         string
+	priorityClassName string
+	plugins           []v1alpha2.Plugin
+	seedJob           *[]v1alpha2.SeedJob
+	groovyScripts     v1alpha2.GroovyScripts
+	casc              v1alpha2.ConfigurationAsCode
+}
+
+func createJenkinsCRFromSample(t *testing.T, sample JenkinsE2EConfiguration) *v1alpha2.Jenkins {
```
Comment (tomaszsek) [inline test/e2e/jenkins.go:73]:
```suggestion
func createJenkinsCRFromConfiguration(t *testing.T, sample JenkinsE2EConfiguration) *v1alpha2.Jenkins {
```

File: test/e2e/jenkins.go [hunk 10]
```diff
@@ -51,12 +51,39 @@ func getJenkinsMasterPod(t *testing.T, jenkins *v1alpha2.Jenkins) *corev1.Pod {
...
+	j := createJenkinsCR(t, sample.name, sample.namespace, sample.seedJob, sample.groovyScripts, sample.casc, sample.priorityClassName)
+	plugins := sample.plugins
+	if len(plugins) > 0 {
+		j.Spec.Master.Plugins = plugins
+	}
```
Comment (tomaszsek) [inline test/e2e/jenkins.go:78]: it won't work because Jenkins CR is already saved without pulgins

File: test/e2e/openshift_test.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		plugins:           plugins,
+		seedJob:           &[]v1alpha2.SeedJob{},
+		groovyScripts:     v1alpha2.GroovyScripts{},
+		casc:              v1alpha2.ConfigurationAsCode{},
+	}
+	jenkins := createJenkinsCRFromSample(t, sample)
+	if !arrayContainsArray(plugins, jenkins.Spec.Master.Plugins) {
+		assert.Fail(t, "Specified plugins not found in the created Jenkins")
+	}
+}
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:73]: move plugins installation to the `TestBaseOpenShiftConfiguration`

File: test/e2e/openshift_test.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+
+func contains(array []v1alpha2.Plugin, element v1alpha2.Plugin) bool {
+	for _, a := range array {
+		if a == element {
+			return true
+		}
+	}
+	return false
+}
```
Comment (tomaszsek) [inline test/e2e/openshift_test.go:91]: not needed

Comment (akram) [reply test/e2e/openshift_test.go:91]: why ?

File: test/e2e/jenkins.go (hunk 10 above)
Comment (akram) [reply test/e2e/jenkins.go]: fixed by doing the assignment before

File: test/e2e/jenkins.go (hunk 9 above)
Comment (akram) [reply test/e2e/jenkins.go]: renamed

File: test/e2e/openshift_test.go (hunk 12 above)
Comment (akram) [reply test/e2e/openshift_test.go:86]: it is a different case to be sure that we can test the when plugins are set and when they are not set (using default)

**Comment by Sig00rd** [comment]
Closing this since the current Operator release doesn't support OpenShift. Thank you for your contribution!

**Pull request closed without merging**